unittests/DefToYAMLConverter/DefToYAMLConverterTest.cpp

```diff
@@ -0,0 +1,90 @@
+//===--- DefToYAMLConverterTest.cpp --------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Localization/LocalizationFormat.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/ADT/StringExtras.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/YAMLParser.h"
+#include "llvm/Support/YAMLTraits.h"
+#include "llvm/Support/raw_ostream.h"
+#include "gtest/gtest.h"
+#include <random>
+
+using namespace swift;
+using namespace swift::diag;
+
+enum LocalDiagID : uint32_t {
+#define DIAG(KIND, ID, Options, Text, Signature) ID,
+#include "swift/AST/DiagnosticsAll.def"
+  NumDiags
+};
+
+static constexpr const char *const diagnosticMessages[] = {
+#define DIAG(KIND, ID, Options, Text, Signature) Text,
+#include "swift/AST/DiagnosticsAll.def"
+};
+
+static std::string getDefaultLocalizationPath() {
+  std::string libPath = llvm::sys::path::parent_path(SWIFTLIB_DIR);
+  llvm::SmallString<128> DefaultDiagnosticMessagesDir(libPath);
+  llvm::sys::path::remove_filename(DefaultDiagnosticMessagesDir); // Remove /lib
+  llvm::sys::path::remove_filename(DefaultDiagnosticMessagesDir); // Remove /.
+  llvm::sys::path::append(DefaultDiagnosticMessagesDir, "share", "swift",
+                          "diagnostics");
+  return std::string(DefaultDiagnosticMessagesDir.str());
+}
+
+/// Random number in [0,n)
+unsigned randNum(unsigned n) { return unsigned(rand()) % n; }
+
+TEST(DefToYAMLConverterTest, missingLocalizationFiles) {
+  ASSERT_TRUE(llvm::sys::fs::exists(getDefaultLocalizationPath()));
+  llvm::SmallString<128> EnglishLocalization(getDefaultLocalizationPath());
+  llvm::sys::path::append(EnglishLocalization, "en");
+  llvm::sys::path::replace_extension(EnglishLocalization, ".yaml");
+  ASSERT_TRUE(llvm::sys::fs::exists(EnglishLocalization));
+  llvm::sys::path::replace_extension(EnglishLocalization, ".db");
+  ASSERT_TRUE(llvm::sys::fs::exists(EnglishLocalization));
+}
+
+TEST(DefToYAMLConverterTest, matchDiagnosticMessagesSequentially) {
+  llvm::SmallString<128> EnglishLocalization(getDefaultLocalizationPath());
+  llvm::sys::path::append(EnglishLocalization, "en");
+  llvm::sys::path::replace_extension(EnglishLocalization, ".yaml");
+  YAMLLocalizationProducer yaml(EnglishLocalization.str());
```

Comment:
I think this test needs to do following:

1. Create a temporary file;
2. use converter to write to that file;
3. use YAMLLocalizationProducer over file populated in step 2
4. Verify that contents of the YAML match what what in .def (which is already done).

Reply:
Here's what I tried:

0. I modified `DefToYAMLConverter` to also accept a filename path.
1. I constructed a path to the `DefToYAMLConverter` executable.
2. Created a temp file.
3. Used the YAMLLocalizationProducer to write to the temp file path.

My code:
```cpp
  llvm::SmallString<128> defYAMLConverter(getDefToYAMLConverterPath());
  defYAMLConverter.append(" --output-filename=");

  llvm::SmallString<128> tempFilename;
  std::error_code EC = llvm::sys::fs::createTemporaryFile("en", "yaml", tempFilename);
  ASSERT_FALSE(EC);

  defYAMLConverter.append(tempFilename);
  std::system(tempFilename.c_str());
  YAMLLocalizationProducer yaml(tempFilename.str());

  // Finally, do random or sequential testing... 
```

The problems are:
1. I don't know whether `std::system` is the right method for this. I tried to see if there is anything like it in `LLVM` but couldn't find anything.
2. When I run the code above I get the following error `sh: /var/folders/69/pr6q8w_j1pvcvr_dxdckwpzr0000gn/T/en-49fe0f.yaml: Permission denied`. How can I make it so I have permission to write to the file?